### DISCUSSION TITLE
airgap images are being GC'd when node is under pressure

### DISCUSSION THREAD

**Discussion by sdemura** [k3s-io/k3s#7570] [Q&A]
**Environmental Info:**
K3s Version: 1.26.4+k3s1

Node(s) CPU architecture, OS, and Version: 

```
# uname -a
Linux lima-ubuntu-lts 5.15.0-67-generic #74-Ubuntu SMP Wed Feb 22 14:14:39 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
```
Cluster Configuration:

single node

**Describe the bug:**

Reopens https://github.com/k3s-io/k3s/issues/3193

Related:
https://github.com/k3s-io/k3s/pull/3755
https://github.com/k3s-io/k3s/issues/2961#issuecomment-781728276
**Steps To Reproduce:**

- Install k3s with airgap archive at `/var/lib/rancher/k3s/agent/images` (logs confirm the file is found and loaded)
- Fill the disk to exceed memory pressure, (I used `xfs_mkfile` for this)
- Watch airgapped images get GC'd

**Expected behavior:**

It's my understanding that images loaded via airgap archive are 'leased', such that kubelet GC will not clean them up if disk thresholds are met.

**Actual behavior:**

Images are GC'd

**Additional context / logs:**

Images before filling disk.
```
IMAGE                                        TAG                    IMAGE ID            SIZE
docker.io/rancher/klipper-helm               v0.7.7-build20230403   48957d5c01a77       248MB
docker.io/rancher/klipper-lb                 v0.4.3                 1c3d6ab0f26a0       9.94MB
docker.io/rancher/local-path-provisioner     v0.0.24                10ada9a7f8ab5       13.9MB
docker.io/rancher/mirrored-coredns-coredns   1.10.1                 97e04611ad434       14.6MB
docker.io/rancher/mirrored-library-busybox   1.34.1                 9bf1b624e59da       3.96MB
docker.io/rancher/mirrored-library-traefik   2.9.4                  28a9c4decb28b       35.6MB
docker.io/rancher/mirrored-metrics-server    v0.6.2                 ea7b05ebc8e8d       26.2MB
docker.io/rancher/mirrored-pause             3.6                    7d46a07936af9       250kB
```

Images after filling disk
```
root@lima-ubuntu-lts:~# crictl image list
IMAGE                                        TAG                 IMAGE ID            SIZE
docker.io/rancher/local-path-provisioner     v0.0.24             10ada9a7f8ab5       13.9MB
docker.io/rancher/mirrored-coredns-coredns   1.10.1              97e04611ad434       14.6MB
docker.io/rancher/mirrored-library-traefik   2.9.4               28a9c4decb28b       35.6MB
docker.io/rancher/mirrored-metrics-server    v0.6.2              ea7b05ebc8e8d       26.2MB
docker.io/rancher/mirrored-pause             3.6                 7d46a07936af9       250kB
```

Containerd log snippet
```
time="2023-05-17T15:40:13.690642637Z" level=info msg="RemoveImage \"sha256:1c3d6ab0f26a050fd411367352b066b867211c94c78360a3e3788df109e4a345\""
time="2023-05-17T15:40:13.691743687Z" level=info msg="ImageDelete event &ImageDelete{Name:docker.io/rancher/klipper-lb:v0.4.3,XXX_unrecognized:[],}"
time="2023-05-17T15:40:13.692913861Z" level=info msg="ImageDelete event &ImageDelete{Name:sha256:1c3d6ab0f26a050fd411367352b066b867211c94c78360a3e3788df109e4a345,XXX_unrecognized:[],}"
time="2023-05-17T15:40:13.693905951Z" level=info msg="RemoveImage \"sha256:1c3d6ab0f26a050fd411367352b066b867211c94c78360a3e3788df109e4a345\" returns successfully"
```

k3s log
```
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/mirrored-metrics-server:v0.6.2"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/mirrored-pause:3.6"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/klipper-helm:v0.7.7-build20230403"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/klipper-lb:v0.4.3"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/local-path-provisioner:v0.0.24"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/mirrored-coredns-coredns:1.10.1"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/mirrored-library-busybox:1.34.1"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported docker.io/rancher/mirrored-library-traefik:2.9.4"
May 17 15:25:15 lima-ubuntu-lts k3s[6117]: time="2023-05-17T15:25:15Z" level=info msg="Imported images from /var/lib/rancher/k3s/agent/images/k3s-airgap-images-arm64.tar.zst in 4.295791625s"
```

Boltbrowser shows the lease exists
<img width="1538" alt="image" src="https://github.com/k3s-io/k3s/assets/11758459/1c55db99-b09e-4c25-91e3-62c03269e496">

**Comment by brandond** [comment]
Unfortunately, it turns out that containerd leases only prevent the layers from getting cleaned up; there is no way to lease lock a tag.

At this point, the best bet is probably to tune the kubelet's GC thresholds, and ensure that your nodes never exceed the configured disk utilization percentages.

**Reply by retr0h** [reply]
Out of curiosity how are leases preserving layers helpful?  When placing an image archive in `/var/lib/rancher/k3s/agent/images`, and having the `imagePullPolicy: Never`, wouldn't the expectation be that everything defined in the archive be preserved?  Especially, when nearly everyone references images by tag in k8s manifets.

From an end users perspective, wouldn't that expectation hold up?

Is there a workaround, aside from increasing disk or GC, as a system that may start to run out of disk space could have images GCed and future jobs unable to run.

**Reply by dMARLAN** [reply]
Is this still true in 1.29? @brandond 

I've tried playing with the `--embedded-registry` or trying to apply pinned labels (`io.cri.containerd.pinned=pinned` and `io.cattle.k3s.pinned=pinned`) but I am still observing airgap/tarball images being GC'd after the pods are evicted due to disk pressure.

I'm not sure what to do to prevent airgap images from being deleted.

**Reply by retr0h** [reply]
@sdemura and I ended up managing it ourselves in our deployment tool.  We turn off the kubelet's GC all together, and manage the images we GC with nerdctl.

**Reply by ashok-mariyala** [reply]
@brandond We faced this issue in 1.29 as well. We have 50% disk available (20GB) in the node still k3s is deleting airgap images. Please let us know what might be the issue is? What are the suggestible GC configurations. Please let us know.

**Reply by brandond** [reply]
If you actually have 50% of your disk space available then the kubelet shouldn't be GCing anything. The default image GC thresholds are 80/85%. The image GC stuff is negotiated between the kubelet and containerd; there is nothing in k3s itself that removes images.

You can check the upstream kubelet documentation for info on GC thresholds:
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

>
> --image-gc-high-threshold int32     Default: 85
> The percent of disk usage after which image garbage collection is always run. Values must be within the range [0, 100], To disable image garbage collection, set to 100.
>  
>--image-gc-low-threshold int32     Default: 80
>The percent of disk usage before which image garbage collection is never run. Lowest disk usage to garbage collect to. Values must be within the range [0, 100] and should not be larger than that of --image-gc-high-threshold.

**Reply by nageswarao-pragada** [reply]
@brandond Thanks for the quick response. I understand issue is not with k3s. What might be the issue? I didn't deleted airgap images manually from containerd. Also, I have 50% Disk available. How do i debug this issue further? Please provide your valuable suggestions. Thanks in advance.

**Reply by brandond** [reply]
You could check the kubelet logs? You might also confirm that you actually have 50% of disk available on the partition holding container images. The kubelet doesn't care if there is 50% free on some OTHER partition, it specifically looks at utilization on the filesystem where the containerd image store lives.
